### DISCUSSION TITLE
Statistics stage

### DISCUSSION THREAD

**Discussion by atherdon** [atherdon/dash-back#60] [Ideas]
While this stage might sound unnecessary, it might become an important stage for the editorial team from a business perspective.
I don't have a clear plan for how we should do it. Plus, there is ""KPI" dashboard that our inner developer is doing when he has free time. 

But he doesn't have it very often, so we might have some advantage from it.

The main idea is to measure the performance of editors and see ups and downs and understand the correlation between them.
This is why I'm trying to generate some basic reports each week.

while it's good - it's still not enough for the team of 8 ppl. plus, soon we will hire few interns - they will add some mess into the process too:)

---

Я добавил пример отчета, который мы сейчас делаем руками в этот таск: https://github.com/atherdon/dash-back/issues/74
Мы также сделали HTML версию, находится она тут(папка dist)
- https://github.com/atherdon/dashb/blob/main/frontend/dist/report-edited.html
- https://github.com/atherdon/dashb/blob/main/frontend/dist/report-published.html

но нам нет смысла фокусироваться только на этом репорте. мы можем делать гораздо более детальные отчеты, используя данные, которые мы накопим в нашей базе. Впринципе, можно будет сделать экспорт данных, которые есть у нас в excel файле, оптимизировать их под базу, которая у нас будет. Таким образом мы сразу будем иметь набор старых данных(с октября) и новых данных. это позволит нам видеть прогресс, который мы делаем как команда.

**ЗАчем нам вообще нужна эта статистика?**

Очень сложно оценить работу каждого редактора, используя только данные - "сколько статей было выложено за неделю".
Статьи бывают простые и сложные, длинные и короткие, и т.д.

Плюс, мы вводим больше "актеров", которые должны влиять на количество обработанных статей.
если раньше редактор делал всю работу сам - теперь есть отдельные люди которые смотрят на внешний вид статьи, и если нужно - они ее исправляют. Аплоадят картинки, переносят код в редактор кода и т.д.

Эти люди упрощают работу редактора, и по логике - он должен выкладывать больше статей.
Но этого не происходит сейчас :)

у нас стабильное число выложенных статей в июле-октябре было 187. при этом, по моим подсчетам - с тем размером команды мы должны были выкладывать около 220 статей, но чтото нам постоянно мешало.

в ноябре в команде добавилось +2 редактора и количество статей при этом не увеличилось особо :) теперь мы реально можем выкладывать 260 статей не перенапрягаясь - но мы застряли на цифре в 200-205.

это или инерция, или нвогодние праздники, или чтото еще - сейчас непонятно. Для этого нам и нужна новая админка, и крутая статистика, которая позволит моему боссу видеть всю информацию на "приборной панели" и делать соответствующие moves.

---

Также, новая статистика и админка должна "поломать старые взгляды" на работу редактора.
дело в том, что сейчас мы publish статьи на текущий день. Тоесть большинство редакторов проснутся в понедельник и начнут выкладывать статьи на понедельник. при этом, в выходные - большинство людей отдыхает, и мало новых статей выкладывается.

Еще мне не нравится, что мы до сих пор мыслим понятием "одна статья". Выкладывая одну статью - я ничего не меняю в глобальном плане. Я хочу чтобы редактора со временем изменили свои единицы измерения, и 1e - было или 5 статей или 10ть.

например - если редактор сел работать, и за один час выложил 5 статей - это хорошо.
если редактор в течении дня выложил 10 и более статей - это тоже хорошо.

это нужно для того, чтобы все редакторы начали думать о big picture и стали работать вместе над целью выкладывать больше статей, при этом не тратя кучу своей энергии.

Потому что иногда получается такая ситуация: редактор всю неделю был занят другими задачами, и выложил только 5 статей за 5 дней. А в воскресенье он садится делать подвиг и выкладывает еще 40 статей на одну тему. Такой подход тоже губителен. проблему он решает только одну(чтобы на следующем собрании наш босс не ругал его). При этом он сильно устает, злится и т.д.
Гораздо проще изменить workflow работы, выкладывать статьи на несколько дней вперед, и даже если он работал над другими задачами - ему не нужно будет сильно дергаться.

---

Нам нужны статы, и много.
Мы можем сохранять очень много информации, которая потом будет анализироваться позже. Особенно если эти данные помогут нам улучшить нашу работу.
может есть какой то модуль, или нам нужно будет создать его, чтобы он покрыл множество cases.
я бы погуглил чтото типо `graphql stats module`.

We can start with simple tracking. How many articles were added, during this(current week)?
Какие типы статей были выложены? нужно ли нам вывод в upload form или всетаки в основной dashboard сделать ссылку, которая будет скрывать published/completed articles?
Делать ли нам статистику после Stage 5? или делать уже после работы над основными entities, как statuses and stages?

We can start with simple calculations, related to uploaded articles, but as it will be a common thing  between different types, probably it should be a separate schema/logic that will be independent.

Часть статистики мы будем выводить - чтобы редакторы в течении недели смогли следить за своим прогрессом. часть статистки можно будет сделать доступной только начальству, которое будет уже делать общий анализ.

**Comment by atherdon** [comment]
@kolserdav я добавил информации, но она очень сумбурная. можете меня скорректировать вашими вопросами?

**Comment by kolserdav** [comment]
Я не стал выделять данные по статистике в отдельную таблицу, так как у нас уже есть связь статей с тегами, как "многие ко многим", также могут появиться зависимые от конкретных типов связи "один к одному" или "один ко многим", в общем не стал делать ещё один к одному - чтобы не городить огород без очевидных причин. С одной стороны, сейчас по статистике, можно просто добавить фильтры по датам операций над статьями и по avg для вывода статей в таблицы и графики. Но всё же отталкиваться лучше  от функционала страницы статистики, тогда будет видно как удобней будет с ней работать и появятся ли смысл в отделении данных статистики в отдельную сущность.

**Reply by kolserdav** [reply]
В этом случае как и во многих других верна проверенная истина. Начинать разработку лучше с UX/UI.
1. Это упрощает сам процесс планирования, так как там простые и понятные термины (кнопка, картинка, список и т.д). 
2. Синхронизирует взаимопонимание фронта и бека.
3. Заставляет всех учитывать нужные моменты и избавляет от "необходимости" делать ненужные конечному продукту вещи.
4. Ускоряет разработку, так как снимает целый пласт вопросов.


Если есть требования к внешнему виду, то он выражен в дизайне, если нет то просто мокап.

А когда сначала делают бек, потом дизайн или вообще без диза, то может оказаться, что в беке не всё рассчитано под конкретный диз, и поэтому он как швейцарский нож умеет много из того, что никогда не будет использовано, но вновь появившиеся неожиданные требования возможно без костыля сходу уже не так просто прикрутить.

**Comment by atherdon** [comment]
Но если я начну заниматься сейчас ux - я застряну надолго. Бекенд хорошо
идёт. Но я понимаю о чем вы говорите. Наверное надо начать с простой
статистики.

**Reply by kolserdav** [reply]
Задание на простую статискику выглядело бы примерно так:
1. На странице статистики отображается таблица с данными статистики работы над статьями.
2. Есть возможность сортировки по колонкам.
3. Таблица строится по пагинации, внизу стрелки переходов по частям выборки и возможность выбрать количество на странице.
4. Вверху есть возможность отфильтровать по редактору.
5. Также вверху есть селект которым можно сгруппировать по редактору и по месяцам. 
И т.д.
Я уже сталкивался, когда бекенд приложения был написан в рассчете на схему, которая осталась в голове предыдущего разработчика, поэтому знаю почему в таких вопросах нужно требовать четкости плана.

**Reply by atherdon** [reply]
я тут полностью с вами согласен. просто я опять пытаюсь усидеть на всех стульях.
и статистику хочется, и нужно понимать что есть еще KPI дашбоард на основном сайте которая то работает то нет, ну и еще есть кучу "хотелок" которые могу помочь в работе.
Согласен и понимаю что это вообще неудобно. 

поэтому я создал подзадачу, связанную со статистикой и в которой у меня есть больше понимания, чего я хочу: https://github.com/atherdon/dash-back/discussions/93
